src/logger.ts

```diff
@@ -15,49 +15,114 @@
+    if (this._recording)
+      this._recording.push(`[${this._name}] ${message}`);
+    this._loggerSink.log(this._name, severity, message, args, this._hints);
+  }
 
-export interface Logger {
-  isEnabled(name: string, severity: LoggerSeverity): boolean;
-  log(name: string, severity: LoggerSeverity, message: string | Error, args: any[], hints: { color?: string }): void;
+  recording(): string[] {
+    return this._recording ? this._recording.slice() : [];
+  }
 }
 
-export const errorLog: Log = { name: 'generic', severity: 'error' };
-export const apiLog: Log = { name: 'api', color: 'cyan' };
+export class Loggers {
+  readonly api: Logger;
+  readonly browser: Logger;
+  readonly protocol: Logger;
+
+  constructor(userSink: LoggerSink | undefined) {
+    const loggerSink = new MultiplexingLoggerSink();
+    if (userSink)
+      loggerSink.add('user', userSink);
+    if (helper.isDebugMode())
+      loggerSink.add('pwdebug', new PwDebugLoggerSink());
+    loggerSink.add('debug', new DebugLoggerSink());
 
-export function logError(logger: InnerLogger): (error: Error) => void {
-  return error => logger.log(errorLog, error, []);
+    this.api = new Logger(loggerSink, 'api', { color: 'cyan' });
+    this.browser = new Logger(loggerSink, 'browser', {});
+    this.protocol = new Logger(loggerSink, 'protocol', { color: 'green' });
+  }
 }
 
-export class InnerLogger {
-  private _userSink: Logger | undefined;
-  private _debugSink: DebugLogger;
+class MultiplexingLoggerSink implements LoggerSink {
+  private _loggers = new Map<string, LoggerSink>();
 
-  constructor(userSink: Logger | undefined) {
-    this._userSink = userSink;
-    this._debugSink = new DebugLogger();
+  add(id: string, logger: LoggerSink) {
+    this._loggers.set(id, logger);
   }
 
-  isLogEnabled(log: Log): boolean {
-    const severity = log.severity || 'info';
-    if (this._userSink && this._userSink.isEnabled(log.name, severity))
-      return true;
-    return this._debugSink.isEnabled(log.name, severity);
+  get(id: string): LoggerSink | undefined {
+    return this._loggers.get(id);
+  }
+
+  remove(id: string) {
+    this._loggers.delete(id);
+  }
+
+  isEnabled(name: string, severity: LoggerSeverity): boolean {
+    for (const logger of this._loggers.values()) {
+      if (logger.isEnabled(name, severity))
+        return true;
+    }
+    return false;
   }
 
-  log(log: Log, message: string | Error, ...args: any[]) {
-    const severity = log.severity || 'info';
-    const hints = log.color ? { color: log.color } : {};
-    if (this._userSink && this._userSink.isEnabled(log.name, severity))
-      this._userSink.log(log.name, severity, message, args, hints);
-    this._debugSink.log(log.name, severity, message, args, hints);
+  log(name: string, severity: LoggerSeverity, message: string | Error, args: any[], hints: { color?: string }) {
+    for (const logger of this._loggers.values())
+      logger.log(name, severity, message, args, hints);
```

Comment:
I think we should check each particular logger's isEnabled() method, otherwise we log to each logger when one of them is enabled.

Reply:
Done.